zrml/parimutuel/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+                payoff_ratio_mul_base.saturated_into(),
+                winning_balance.saturated_into(),
+            )?
+            .saturated_into();
+
+            Self::check_values(
+                winning_balance,
+                pot_total,
+                outcome_total,
+                payoff_ratio_mul_base,
+                payoff,
+            )?;
+
+            let slashable_asset_balance = winning_balance;
+
+            T::AssetManager::slash(winning_asset, &who, slashable_asset_balance);
+
+            let remaining_bal = T::AssetManager::free_balance(market.base_asset, &pot_account);
+            let actual_payoff = payoff.min(remaining_bal);
+
+            T::AssetManager::transfer(market.base_asset, &pot_account, &who, actual_payoff)?;
+
+            Self::deposit_event(Event::RewardsClaimed {
+                market_id,
+                asset: winning_asset,
+                balance: slashable_asset_balance,
+                actual_payoff,
+                sender: who.clone(),
+            });
+
+            Ok(())
+        }
+
+        fn do_claim_refunds(who: T::AccountId, refund_asset: AssetOf<T>) -> DispatchResult {
+            let market_id = match refund_asset {
+                Asset::ParimutuelShare(market_id, _) => market_id,
+                _ => return Err(Error::<T>::NotParimutuelOutcome.into()),
+            };
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Resolved, Error::<T>::MarketIsNotResolvedYet);
+            ensure!(market.scoring_rule == ScoringRule::Parimutuel, Error::<T>::InvalidScoringRule);
+            ensure!(
+                matches!(market.market_type, MarketType::Categorical(_)),
+                Error::<T>::NotCategorical
+            );
+            Self::market_assets_contains(&market, &refund_asset)?;
+            let winning_outcome = market.resolved_outcome.ok_or(Error::<T>::NoResolvedOutcome)?;
+            let pot_account = Self::pot_account(market_id);
+            let winning_asset = match winning_outcome {
+                OutcomeReport::Categorical(category_index) => {
+                    Asset::ParimutuelShare(market_id, category_index)
+                }
+                OutcomeReport::Scalar(_) => return Err(Error::<T>::NotCategorical.into()),
+            };
+            let outcome_total = T::AssetManager::total_issuance(winning_asset);
+            ensure!(outcome_total == <BalanceOf<T>>::zero(), Error::<T>::RefundNotAllowed);
+
+            let refund_balance = T::AssetManager::free_balance(refund_asset, &who);
+            ensure!(!refund_balance.is_zero(), Error::<T>::RefundableBalanceIsZero);
+            debug_assert!(
+                refund_asset != winning_asset,
+                "Since we were checking the total issuance of the winning asset to be zero, if \
+                 the refund balance is non-zero, then the winning asset can't be the refund asset!"
+            );
```

Comment:
No logs in optimized builds.

Reply:
d49ac0b28df75792f2adfd8403409bc439b164de